### ISSUE TITLE
contrib/go-redis/redis: WithContext causes a random spike in connections and timeouts

### ISSUE THREAD

**Issue by goodspark** [DataDog/dd-trace-go#608]
go-redis v6.15.7
Redis (on AWS ElastiCache) v3.2.4-compatible engine
DD trace v1.22.0

We've noticed periodic spikes in connections and timeouts to redis when we use DD's tracing. We haven't had a chance to investigate more, but we did notice those spikes went away when tracing was removed. For now I'm opening an issue to track it.

DD APM:

![Screenshot from 2020-03-09 13-28-54](https://user-images.githubusercontent.com/29210237/76254683-21885800-620a-11ea-87ea-f653d1beec33.png)

Notice the two spikes seen in p95 latency on Wed and Thur in the image above. Since then, no such issues.

AWS CloudWatch metrics for the impacted redis instance:

![image (1)](https://user-images.githubusercontent.com/29210237/76255250-1da90580-620b-11ea-9935-e025c59230f9.png)

Our code was more or less doing:

```go
// During server startup
client := redistrace.NewClient(&redis.Options{
  Addr: "redis endpoint",
})

// In our request handlers/code
stuff := client.WithContext(ctx).Get("key").Result()
```

**Comment by gbbr** [comment]
You're saying that:

> WithContext is causes a random spike [...]

Are you saying that if you remove `WithContext` the spike is gone? That would help narrow down the issue.

**Comment by goodspark** [comment]
Yes

**Comment by gbbr** [comment]
Thanks for confirming. It's probably the change introduced in https://github.com/DataDog/dd-trace-go/pull/530. We'll look into it.

**Comment by tanordheim** [comment]
I think I'm observing the same problem here just manifested differently. I don't see any latency spikes per se - all though the service I'm having the issue in is very low traffic right now so I probably wouldn't notice spikes even if there were any.

What I'm noticing however is that the DataDog tracer starts logging this after the service has handled a few requests:

```
2020/03/20 13:19:04 Datadog Tracer v1.22.0 ERROR: payload queue full, dropping 1 traces, 200+ additional messages skipped (first occurrence: 20 Mar 20 13:18 UTC)
2020/03/20 13:18:14 Datadog Tracer v1.22.0 ERROR: lost 4344 traces: Post "http://10.43.3.197:8126/v0.4/traces": context deadline exceeded (Client.Timeout exceeded while awaiting headers), 2 additional messages skipped (first occurrence: 20 Mar 20 13:17 UTC)
```

The service does one Redis `GET` for each HTTP request handled, however I see in the DataDog traces that while the service handles around 80 requests per minute I receive tens of thousands of Redis traces within the same time window.

Redis traces:

![image](https://user-images.githubusercontent.com/49933/77167638-b286eb00-6ab6-11ea-994d-d35e3e1899e4.png)

HTTP request traces:

![image](https://user-images.githubusercontent.com/49933/77167657-badf2600-6ab6-11ea-8b24-832fc3840a6f.png)

The way my code was written was something to the effect of this:

```go
type Svc struct {
  redis *redis.Client
}

func (s *Svc) DoSomething(ctx context.Context) {
  client := ddredis.WrapClient(s.redis).WithContext(ctx)
  client.Get("...")
}
```

Then my HTTP handler in effect called this with `svc.DoSomething(r.Context())`.

By changing my code to no longer use `WithContext`, and by just storing the wrapped client in the `Svc` directly instead of wrapping it in each function call, the number of Redis traces received went down to match the number of HTTP requests served.

I don't know if this was caused by me wrapping the Redis client in each method call and then calling `WithContext`, or if it's because of the `WithContext` call itself - but this still feels like incorrect behavior somehow and something that caused our service to go down in production as Kubernetes started OOMKilling it as it leaked memory over time; presumably from unflushed traces or something like that judging from the log messages.

Redis traces after changing the code:

![image](https://user-images.githubusercontent.com/49933/77168504-3c838380-6ab8-11ea-8864-e8e8f8ce03e3.png)

HTTP traces after changing the code:

![image](https://user-images.githubusercontent.com/49933/77168531-44432800-6ab8-11ea-86d2-55428e7d718f.png)

**Comment by knusbaum** [comment]
The issue here is that `ddredis.WrapClient` modifies the redis client itself by calling `WrapProcess`. This lets us add some middleware that creates a span for each redis call:
https://github.com/DataDog/dd-trace-go/blob/389586928d8333724f68a57b80607be23517d338/contrib/go-redis/redis/redis.go#L80

Each time you call `WrapClient`, a new wrapper is added, and an additional span will be created for every redis call. I was able to reproduce this with the following code:
```
package main

import (
	"context"
	"fmt"

	"github.com/go-redis/redis"
	ddredis "gopkg.in/DataDog/dd-trace-go.v1/contrib/go-redis/redis"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
)

func tryRedis(c *redis.Client) {
	ctx := context.Background()
	client := ddredis.WrapClient(c).WithContext(ctx)
	for i := 0; i < 10; i++ {
		val, err := client.Get("key").Result()
		if err != nil {
			fmt.Printf("Error: %s\n", err)
			return
		}
		fmt.Println("key", val)
	}
}

func main() {
	tracer.Start(tracer.WithServiceName("goredis-test"))
	defer tracer.Stop()

	client := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password set
		DB:       0,  // use default DB
	})

	err := client.Set("key", "value", 0).Err()
	if err != nil {
		panic(err)
	}
	tryRedis(client)
}
```

### One call to `WrapClient`
Here is multiple calls to redis after wrapping the client once:
```
func tryRedis(c *redis.Client) {
	ctx := context.Background()
	client := ddredis.WrapClient(c).WithContext(ctx)
	for i := 0; i < 10; i++ {
		val, err := client.Get("key").Result()
		if err != nil {
			fmt.Printf("Error: %s\n", err)
			return
		}
		fmt.Println("key", val)
	}
}
```
<img width="185" alt="Screen Shot 2020-03-20 at 1 26 24 PM" src="https://user-images.githubusercontent.com/1819836/77195564-d3980d80-6aaf-11ea-8b34-95f46085ceb2.png">
There are 10 calls, and 10 traces show up, as we expect

### Multiple calls to `WrapClient`
Here is one where the client is wrapped before each call:
```
func tryRedis(c *redis.Client) {
	ctx := context.Background()
	for i := 0; i < 10; i++ {
	client := ddredis.WrapClient(c).WithContext(ctx)
		val, err := client.Get("key").Result()
		if err != nil {
			fmt.Printf("Error: %s\n", err)
			return
		}
		fmt.Println("key", val)
	}
}
```
<img width="181" alt="Screen Shot 2020-03-20 at 1 26 30 PM" src="https://user-images.githubusercontent.com/1819836/77195697-09d58d00-6ab0-11ea-8b69-8af6d843ed22.png">
Here, at each iteration of the loop, we `WrapClient`, and so at each iteration through the loop, we will send one more span. First iteration = 1 span, second iteration = 2 spans, etc. 
As expected if `WrapClient` continues to add process wrappers, we end up with 55 spans (sum(1..10))

### Multiple calls to `WithContext`
Here is one that wraps client only once, but creates a context for every iteration:
```
func tryRedis(c *redis.Client) {
	ctx := context.Background()
	client := ddredis.WrapClient(c)
	for i := 0; i < 10; i++ {
		client := client.WithContext(ctx)
		val, err := client.Get("key").Result()
		if err != nil {
			fmt.Printf("Error: %s\n", err)
			return
		}
		fmt.Println("key", val)
	}
}
```
<img width="171" alt="Screen Shot 2020-03-20 at 1 26 37 PM" src="https://user-images.githubusercontent.com/1819836/77196255-f4149780-6ab0-11ea-8957-97c3a643f537.png">
It works as we would want it to, only creating 1 span per iteration.

## Solution
The quick solution is not to wrap the redis client multiple times. That's not how we intend the integration to work. In general, initializing integrations is not something you want to do more frequently than you have to, since there is overhead involved that you probably don't want to absorb on every redis call anyway. If there's a reason you don't want to, or can't hold onto the wrapped client, please let us know. We should address that.

We *should* see what we can do to mitigate this issue with `WrapClient`. I'm not sure if it's feasible to check if we've already wrapped the client.

I think we can't just look at the `process` function wrapping because:
1. "Slice, map, and *function* values are not comparable." (https://golang.org/ref/spec#Comparison_operators)
2. If someone else calls `WrapProcess` on the client after we do, the fact that our function is in the process stack is hidden.

Maybe we can deprecate the `WrapClient` function if we think it can't be used safely. We do have `NewClient` which will correctly instantiate and return a wrapped client. However, the `*redis.Client` is accessible anyway through our [wrapped `Client`](https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/contrib/go-redis/redis#Client).

We may also modify the integration so it doesn't use `WrapProcess`, but I'm not sure how easy it is to trace `go-redis` without using that.

A simple option is to add a warning to the documentation. I'm not sure how common this pattern is or how many people have hit it. The fact that it blows up the tracer makes me want to mitigate with something more than a warning, though.

**Comment by goodspark** [comment]
We're not calling WrapClient multiple times. See original description. Your comment may be applicable to Trond, however.

**Comment by goodspark** [comment]
We're not repeatedly wrapping a redis client. Your comment may be applicable to Trond, however. But it does not address the issue.

**Comment by knusbaum** [comment]
Thank you for the info, @goodspark. There may be two different issues here.

**Comment by knusbaum** [comment]
@goodspark 
I ran some benchmarks on the integration against a redis instance running locally.
```
// Untraced
func BenchmarkRedisCalls(b *testing.B) {
	tracer.Start(tracer.WithServiceName("goredis-test"))
	defer tracer.Stop()

	client := SetupRedis()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, err := client.Get("key").Result()
		if err != nil {
			panic(err)
		}
	}
}
```
```
// Traced
func BenchmarkRedisCalls(b *testing.B) {
	tracer.Start(tracer.WithServiceName("goredis-test"))
	defer tracer.Stop()

	client := ddredis.WrapClient(SetupRedis())
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, err := client.Get("key").Result()
		if err != nil {
			panic(err)
		}
	}
}
```
```
// Traced + Context
func BenchmarkRedisCalls(b *testing.B) {
	tracer.Start(tracer.WithServiceName("goredis-test"))
	defer tracer.Stop()

	ctx := context.Background()
	client := ddredis.WrapClient(SetupRedis())
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		client := client.WithContext(ctx)
		_, err := client.Get("key").Result()
		if err != nil {
			panic(err)
		}
	}
}
```

I ran them with:
```
go test -run NONE -bench 'BenchmarkRedisCalls' -benchmem -benchtime 10s -count 10 | tee untraced.txt 
go test -run NONE -bench 'BenchmarkRedisCalls' -benchmem -benchtime 10s -count 10 | tee traced.txt 
go test -run NONE -bench 'BenchmarkRedisCalls' -benchmem -benchtime 10s -count 10 | tee traced_with_context.txt
```
Then compared with `benchstat` to get the following results:
```
benchstat untraced.txt traced.txt traced_with_context.txt 
name \ time/op    untraced.txt  traced.txt  traced_with_context.txt
RedisCalls-8        873µs ± 8%  996µs ±12%              1018µs ±10%

name \ alloc/op   untraced.txt  traced.txt  traced_with_context.txt
RedisCalls-8         184B ± 0%  3508B ± 6%               3672B ± 6%

name \ allocs/op  untraced.txt  traced.txt  traced_with_context.txt
RedisCalls-8         6.00 ± 0%  48.00 ± 0%               53.00 ± 0%
```

Based on these, there isn't significant time overhead with the traced client (~100-200µs) per call.
There is quite a bit more memory allocated with the traced client (200 bytes compared to ~3.5k). We can look more closely at what is being allocated, but I'm not sure if that can account for the latency you're seeing if the time/op isn't significantly higher.

What's more is that the difference between the traced client and the traced client using `WithContext` is quite small, both in memory and time.

These are just small benchmarks and can't account for everything that would be going on in a real-world application. Can you elaborate on the kind of latency spikes you're seeing? How much latency is there? If you are seeing errors, what kind, etc.

**Comment by goodspark** [comment]
I think the core problem is not latency. Our current guess is `WithContext` is somehow inappropriately using and/or releasing connections, which causes connection saturation, which then causes delays in connection acquisition times for new requests. See the charts in the original post.

We saw spikes in connections to redis around peak traffic hours. Our traces around the redis calls took up to 1s in error cases. The service in question has specific deadline (100-300ms - I forget the exact number). This then causes deadline errors downstream (eg. in a DB call). When we removed the tracing, our connection metrics in AWS returned to normal and our latencies smoothed out and went back to normal.

A synchronous benchmark may not be enough to trigger the problem.

**Comment by tanordheim** [comment]
@knusbaum Yeah I realized after posting the comment and digging a little in the dd-trace-go source code that my specific problem was me mis-using the package API by wrapping our Redis client every time we did a query in our application instead of reusing an already wrapped one (basically combing wrapping and `WithContext` instead of relying just on `WithContext`). As such it likely isn't related to the original issue posted by @goodspark here.

That said, I feel like it's a fairly bad outcome of this API mis-use that at first glance wasn't obvious to me that stemmed from our implementation - is it OK if I leave my comments in place on this issue even if not directly related to the original issue for information purposes in case someone else has the same issue?

**Comment by knusbaum** [comment]
@tanordheim You can leave the comments here, but I've opened a new issue for the problem you found (#619). Please have a look and feel free to add anything I missed.

**Comment by knusbaum** [comment]
@goodspark Thanks for the info. I'll look into that.

**Comment by knusbaum** [comment]
@goodspark 
Earlier you said that when you remove the call to `WithContext` it fixes the issue, and recently you also said:
> When we removed the tracing, our connection metrics in AWS returned to normal and our latencies smoothed out and went back to normal.

Can you confirm that leaving tracing *on* but *not* calling `WithContext` fixes the issue?

I'm working on trying to reproduce the issue now, but want to make sure I'm looking for the right thing (right now comparing a traced client without calling `WithContext` to a traced client that calls `WithContext` for each operation)

**Comment by goodspark** [comment]
Yup. That was a bit poorly worded. We're still using a client created by DD tracing's redis package. We're just not using WithContext anymore, which more or less turns off distributed tracing in golang.

**Comment by knusbaum** [comment]
Thank you and sorry to ask again. Just want to make sure I'm looking at the right things.

**Comment by knusbaum** [comment]
@goodspark 
I ran manual tests for 10 minutes each of a traced redis client, and a traced redis client calling `WithContext` before each call to `Get`.

Code for the run calling `WithContext`. For the run without `WithContext`, the calls to `WithContext` were removed:
```go
package main

import (
	"context"
	"strconv"
	"sync"
	"time"

	"github.com/go-redis/redis"
	ddredis "gopkg.in/DataDog/dd-trace-go.v1/contrib/go-redis/redis"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
)

func tryRedis(c *redis.Client) {
	ctx := context.Background()
	client := ddredis.WrapClient(c)

	stop := make(chan struct{})

	var wg sync.WaitGroup
	wg.Add(1000)
	for i := 0; i < 1000; i++ {
		go func() {
			defer wg.Done()
			for j := 0; j < 1000; j++ {
				k := "key" + strconv.Itoa(i) + strconv.Itoa(j)
				client := client.WithContext(ctx)
				_, err := client.Set(k, "VAL"+strconv.Itoa(i)+strconv.Itoa(j), 0).Result()
				if err != nil {
					panic(err)
					return
				}
			}
			for {
				for j := 0; j < 1000; j++ {
					k := "key" + strconv.Itoa(i) + strconv.Itoa(j)
					client := client.WithContext(ctx)
					_, err := client.Get(k).Result()
					if err != nil {
						panic(err)
						return
					}
				}
				select {
				case <-stop:
					return
				default:
				}
			}
		}()
	}
	time.Sleep(10 * time.Minute)
	close(stop)
	wg.Wait()
}

func main() {
	tracer.Start(tracer.WithServiceName("goredis-test"))
	defer tracer.Stop()

	client := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password set
		DB:       0,  // use default DB
	})
	tryRedis(client)
}
```

Results were uninspiring.
Both showed latency spikes in the max, but stayed pretty level for p95 latency and below.

With `WithContext`:
<img width="395" alt="Screen Shot 2020-03-24 at 11 35 47 PM" src="https://user-images.githubusercontent.com/1819836/78281489-5d000480-74e0-11ea-990e-f5f05f5d0cb1.png">


Without `WithContext`:
<img width="392" alt="Screen Shot 2020-03-24 at 11 36 01 PM" src="https://user-images.githubusercontent.com/1819836/78281513-67ba9980-74e0-11ea-9a8d-0c75435993d9.png">


The average spikes may be slightly better without `WithContext`, with most of them seeming to be ~125ms, whereas with `WithContext`, they seem to bounce to ~150ms or ~175ms. However, the worst max latency was recorded without `WithContext`, so whatever timeouts might be seen with `WithContext` would likely also be seen without it. None of this seems to indicate a significant difference between the two, similar to the benchmark I ran.

My test case doesn’t seem to reproduce the issue. If you can give us more information that might help us reproduce, or even better provide a testcase that reproduces your issue, that will help us move forward on this.

**Comment by knusbaum** [comment]
@goodspark I've been unable to reproduce this issue.
I recommend reaching out to our support team to help handle this issue if you still need support in identifying the cause of this latency.

**Comment by valer-cara** [comment]
I see you're using `go-redis`. I've also encountered issues with NewConnections spiking in Redis/Elasticache.

So far I've been able to track it down to the way go-redis handles it's internal pool. If during a request an error happens, the client decides to [remove](https://github.com/go-redis/redis/blob/9cfca5827def4410970428388a51343632f9396e/redis.go#L276) that connection from the pool. On the next request it'll try to aquire a new one. That's the basic mechanic triggering the spike.

In my case the error is `context.DeadlineExceeded`, since I want low latency, fail-open behavior. The problem is that there are **no bounds** to the number of close/new-conns the library does. The number of conns always stays under PoolSize because 
connections are Closed, not Put back.

Under load, contexts may timeout with a higher probability (scheduler wait, contention/sync blocks).

I'm still trying to figure out a way around this.. One option is to increase the pool size to lower contention in `[waitTurn](https://github.com/go-redis/redis/blob/9cfca5827def4410970428388a51343632f9396e/internal/pool/pool.go#L234)` and lower the odds of contexts timing out. Another is to use the [Limiter](https://github.com/go-redis/redis/blob/6e4eb2e3acad9575fb1e2c8690a3c1decf2e59e5/redis.go#L188) option in `redis.Options`.
